Skip articles in KeywordTitle assessment

nataliashitova opened the pull request:
## Summary

This PR can be summarized in the following changelog entry:

* Skips articles in the beginning of the SEO title when checking if the keyword is present in (the beginning of) the title.
* Checks for the keyword in the title only after a title has been set.

## Relevant technical choices:

* The assessment now skips all articles specified for a given locale when assessing if a keyword is present in the beginning of the title. This means that if a user has a title "A dog" and wants to rank for "dog" the assessment will return a green "good" score.
* This also means that if a user has a title "A the dog" s/he will receive a high score.
* An additional requirement was added to the paper to receive this assessment: The paper has to have a title. It was done to prevent situations, when a user receives a feedback "Please create an SEO title." from TitleWidth assessment and then "The focus keyword '%1$s' does not appear in the SEO title." from KeywordTitle assessment.
* Because of that additional requirement, I had to edit the specs.
* Added an inverted question mark to the word boundary processing and trimming punctuation to account for cases like title: "¿Cómo hacer guacamole?" ranking for "cómo". As a side effect this also fixed the bug described in #1003.

## Test instructions

This PR can be tested by following these steps:

### Browserified

- check out the branch
- use the browserfied example. Don't forget to run a grunt build:js
- set locale to any language or leave default
- add a keyword and a title that starts directly with that keyword - the assessment should return a green bullet "The SEO title contains the focus keyword, at the beginning which is considered to improve rankings."
- add an article in the beginning of your title (e.g., "a", "an", "the" for English) - the assessment result should remain the same
- add a different non-article word in the beginning of your title - the assessment should return an orange bullet "The SEO title contains the focus keyword, but it does not appear at the beginning; try and move it to the beginning."
- remove the keyword from the title - the assessment should return a red bullet "The focus keyword '%1$s' does not appear in the SEO title."

### Wordpress

- Check out `stories/9630-add-titleKeywordAssessment-to-SEOAssessor` in wordpress-seo.
- Run yarn install in the plugin's directory.
- Run yarn add "git+https://github.com/Yoast/yoastseo.js#stories/1313-improve-keyword-title-assessment" in the plugin's directory.
- Navigate to node_modules/yoastseo and run yarn install and grunt build:js.
- Navigate back to the plugin's directory and run grunt build:js.
- Test the same functionality as described in the test instructions for Browserified.

Alternatively, you can use the beta script to test the same combination of branches.

Fixes #1313, and fixes #1003

manuelaugustin commented on src/assessments/seo/titleKeywordAssessment.js:
```diff
@@ -1,46 +1,113 @@
+	 *
+	 * @returns {Object} object with score and text.
+	 */
+	calculateScore() {
```
calculateResult? (in line with the previous assessments that have been altered like this)

nataliashitova replied:
Agreed.

manuelaugustin commented on src/assessments/seo/titleKeywordAssessment.js:
```diff
@@ -1,46 +1,113 @@
+		if ( requiresKeyword ) {
+			return i18n.sprintf( i18n.dgettext(
+				"js-text-analysis",
+				/* Translators: %1$d expands to the keyphrase. */
```
I'm not entirely sure how these translator comments are parsed, but I suspect that now they'll also get this comment in the cases when this variable isn't present (good & okay). Perhaps we should check whether that's an issue.

manuelaugustin commented on src/researches/italian/functionWords.js:
```diff
@@ -245,6 +245,9 @@ module.exports = function() {
 			indefinitePronouns, correlativeConjunctions, subordinatingConjunctions, interrogativeDeterminers, interrogativePronouns,
 			interrogativeAdverbs, locativeAdverbs, miscellaneous, pronominalAdverbs, recipeWords, timeWords, vagueNouns ),
 
+		// Articles are used in the keywordTitle assessment
```
End sentence with period.

manuelaugustin commented on src/researches/portuguese/functionWords.js:
```diff
@@ -167,6 +167,9 @@ module.exports = function() {
 			interrogativeProAdverbs, locativeAdverbs, otherAuxiliaries, copula, subordinatingConjunctions, interviewVerbs, additionalTransitionWords,
 			delexicalizedVerbs, interjections, recipeWords, timeWords, vagueNouns, miscellaneous, titles ),
 
+		// Articles are used in the keywordTitle assessment
```
End sentence with period.

manuelaugustin commented on src/researches/spanish/functionWords.js:
```diff
@@ -290,6 +290,9 @@ module.exports = function() {
 		// These word categories cannot intervene between an auxiliary and a corresponding passive participle.
 		cannotBeBetweenPassiveAuxiliaryAndParticiple: [].concat( copulaEstar, copulaEstarInfinitive ),
 
+		// Articles are used in the keywordTitle assessment
```
End sentence with period.

manuelaugustin commented:
CR 🚧

* Some minor documentation style issues, see comments.

* Code coverage for findKeywordInPageTitle.js has decreased

* I was thinking again about the changelog description. I think it would be good to mention that the titleKeyword assessment is now only shown when there's a title in the changelog, since it's a user-facing change. So something like "The keyword in the title is only checked after a title has been set."

* Wasn't there some issue about the Spanish ¿ that is fixed now? Because then you can also add that to the fixed issues.

nataliashitova commented on src/assessments/seo/titleKeywordAssessment.js:
```diff
@@ -1,46 +1,113 @@
 
-	return assessmentResult;
-};
+		return i18n.dgettext( "js-text-analysis", resultText );
```
Does not seem to matter (and it used to be without a sprinf wrapper in the previous stable version of the assessment), but maybe indeed a safer way.

nataliashitova commented:
@manuelaugustin 

> Some minor documentation style issues, see comments.
Fixed

> Code coverage for findKeywordInPageTitle.js has decreased
I am not seeing it. I checked build print-out and coverage/lcov-report/index.html, both the previous version and the current version give 100% coverage.

> I was thinking again about the changelog description. I think it would be good to mention that the titleKeyword assessment is now only shown when there's a title in the changelog, since it's a user-facing change. So something like "The keyword in the title is only checked after a title has been set."
Fixed

> Wasn't there some issue about the Spanish ¿ that is fixed now? Because then you can also add that to the fixed issues.
I was citing that issue in the description of the pull request. Now I also put it in `Fixes`

> There seem to be some merge conflicts which need to be resolved.
Resolved

manuelaugustin commented:
CR 🚧

* See the screenshot of the code coverage of `findKeywordInPageTitle.js` as I see it in the report (the implicit else condition isn't covered).

* Github seems to be confused again: recent changes in develop (Flesch reading ease, PR about the keywords consisting only of special characters) are also in this PR. If all the branches have been pulled/merged/pushed correctly, probably  just changing the base branch to develop and changing it back again to feature/recalibration should fix the issue.

* There are some merge conflicts.

* I changed the `fixes` in the issue description; you need to add `fixes` before every issue number for Github to recognize it (see https://help.github.com/articles/closing-issues-using-keywords/).

<img width="750" alt="screen shot 2018-05-07 at 11 26 32" src="https://user-images.githubusercontent.com/27805437/39696226-60ee4592-51ed-11e8-8155-27287d7a11e6.png">

nataliashitova commented:
@manuelaugustin 
> See the screenshot of the code coverage of findKeywordInPageTitle.js as I see it in the report (the implicit else condition isn't covered).

Looks like my coverage test was not updating properly, so I could not see the decrease. Now it is fixed and I added a spec to cover the implicit-else.

> Github seems to be confused again: recent changes in develop (Flesch reading ease, PR about the keywords consisting only of special characters) are also in this PR. If all the branches have been pulled/merged/pushed correctly, probably just changing the base branch to develop and changing it back again to feature/recalibration should fix the issue.

Fixed.

> There are some merge conflicts.

Fixed

> I changed the fixes in the issue description; you need to add fixes before every issue number for Github to recognize it (see https://help.github.com/articles/closing-issues-using-keywords/).

Thanks!

jcomack commented on src/researches/findKeywordInPageTitle.js:
```diff
@@ -1,24 +1,42 @@
+	keyword = keyword.toLocaleLowerCase();
+	result.position = title.indexOf( keyword );
+
+	if ( result.matches > 0 && result.position > 0 ) {
```
I'd consider rewriting this so it's an early return:

```
if  ( result.matches === 0 || result.position <= 0 ) {
    return;
}
```

jcomack commented on src/researches/findKeywordInPageTitle.js:
```diff
@@ -1,24 +1,42 @@
-	var result = { matches: 0, position: -1 };
+	let title = removePunctuation( paper.getTitle() );
+	let keyword = escapeRegExp( paper.getKeyword() );
+	const locale = paper.getLocale();
```
Seeing as this variable is only used once on line 26, I'd suggest removing this line and just call `paper.getLocale()` on line 26 as the third parameter.

nataliashitova replied:
also line 43

jcomack commented on src/researches/findKeywordInPageTitle.js:
```diff
@@ -1,24 +1,42 @@
+	result.position = title.indexOf( keyword );
+
+	if ( result.matches > 0 && result.position > 0 ) {
+		const articles = functionWords[ getLanguage( locale ) ].articles;
```
I'd suggest adding more documentation as to _what_ is happening here. Without prior knowledge, it's unclear that you're trying to deal with cases where a function word might exist before the actual keyword.

Dieterrr commented:
Acceptance done 🥇
